runtime/compiler/trj9/optimizer/JProfilingRecompLoopTest.cpp

```diff
@@ -0,0 +1,214 @@
+      TR::Node *node = asyncCheckTreeTop->getNode();
+      int32_t depth = testLocation->second;
+      if (trace())
+         traceMsg(comp, "block_%d, n%dn, depth = %d\n",asyncCheckTreeTop->getEnclosingBlock()->getNumber(), asyncCheckTreeTop->getNode()->getGlobalIndex(), depth);
+      TR_ByteCodeInfo bci = asyncCheckTreeTop->getNode()->getByteCodeInfo();
+      performTransformation(comp, "%s Add recompilation test after asyncCheck node n%dn\n", optDetailString(), node->getGlobalIndex());
+      TR::Node *root = generateBlockFrequecnyCalculationNode(comp, node, bfi, bfi->getOriginalBlockNumberToGetLiveFrequency(bci,comp));
+      
+      TR::Block *originalBlock = asyncCheckTreeTop->getEnclosingBlock();
+      TR::Block *newSplittedBlock = originalBlock->split(asyncCheckTreeTop->getNextTreeTop(), cfg, true, true);
+      TR::Block *callRecompileBlock = TR::Block::createEmptyBlock(node, comp, UNKNOWN_COLD_BLOCK_COUNT);
+      callRecompileBlock->setIsCold(true);
+      
+      TR::Node *callNode = TR::Node::createWithSymRef(node, TR::icall, 2, comp->getSymRefTab()->findOrCreateRuntimeHelper(TR_jitRetranslateCallerWithPrep, false, false, true));
+      callNode->setAndIncChild(0, TR::Node::createWithSymRef(node, TR::loadaddr, 0, comp->getSymRefTab()->findOrCreateStartPCSymbolRef()));
+      callNode->setAndIncChild(1, TR::Node::createWithSymRef(node, TR::loadaddr, 0, comp->getSymRefTab()->findOrCreateCompiledMethodSymbolRef()));
+      TR::TreeTop *callTree = TR::TreeTop::create(comp, TR::Node::create(node, TR::treetop, 1, callNode));
+      callTree->getNode()->setIsProfilingCode();
+      callRecompileBlock->append(callTree);
+      cfg->addNode(callRecompileBlock);
+      
+      TR::Node *loadBaseThreshold = TR::Node::createWithSymRef(node, TR::iload, 0, comp->getSymRefTab()->createKnownStaticDataSymbolRef(&recompileThreshold, TR::Int32));
+      TR::Node *adjustThreshold = TR::Node::create(node, TR::ishl, 2, loadBaseThreshold, TR::Node::iconst(node, (depth-1)*3));
+      TR::Node *cmpNode = TR::Node::createif(TR::ificmple, root, adjustThreshold, newSplittedBlock->getEntry());
+      
+      TR::TreeTop *cmpFlag = TR::TreeTop::create(comp, asyncCheckTreeTop, cmpNode);
+      newSplittedBlock->getEntry()->insertTreeTopsBeforeMe(callRecompileBlock->getEntry(), callRecompileBlock->getExit());
+      cfg->addEdge(TR::CFGEdge::createEdge(callRecompileBlock, newSplittedBlock, comp->trMemory()));
+      cfg->addEdge(TR::CFGEdge::createEdge(originalBlock, callRecompileBlock, comp->trMemory()));
+      if (trace())
+         traceMsg(comp,"\t\t Newly created recompilation Test : n%dn in block_%d\n/\t\tRecompilation Call in block_%d\n",
+            callNode->getGlobalIndex(), originalBlock->getNumber(), callRecompileBlock->getNumber());
+      }
+   }
+
+TR::Node *
+TR_JProfilingRecompLoopTest::generateBlockFrequecnyCalculationNode(TR::Compilation *comp, TR::Node *node, TR_BlockFrequencyInfo *bfi, int32_t blockNumber)
```

Comment:
There is code duplication here between this opt and the where we would insert the recomp tree at the start of the method - can you please make sure to not duplicate this code? We need to common the code out and parameterize the tree generation.